### PR TITLE
fix(docker): sharing dependency cache

### PR DESCRIPTION
**Pull request by aarnphm**
https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/syntax.md#run---mounttypecache

Currently, dependency cache will be invalid for any subsequent `bentoml build` and `bentoml containerize`.
This PR address this by assigning the checksum of given bentofile.yaml as a cache key.

This improves build time for subsequent iteration by 40%

### CODE REVIEW

File: bentoml/_internal/bento/bento.py [hunk 1]
```diff
@@ -138,8 +138,9 @@ def info(self) -> BentoInfo:
     def create(
         cls,
         build_config: BentoBuildConfig,
-        version: t.Optional[str] = None,
-        build_ctx: t.Optional[str] = None,
+        version: str | None = None,
+        build_ctx: str | None = None,
+        hash_key: str = "",
```
Comment (sauyon) [inline bentoml/_internal/bento/bento.py:143]: Can we just hash the relevant parts of `build_config` here instead?

File: bentoml/_internal/bento/docker/templates/_macros.j2 [hunk 2]
```diff
@@ -39,3 +43,4 @@ if [ -f {{ path }} ]; then
 fi
 EOF
 {%- endmacro -%}
+
```
Comment (sauyon) [inline bentoml/_internal/bento/docker/templates/_macros.j2:46]: extraneous newline?

**Review by sauyon — APPROVED** [review]
LGTM minus one tiny niggle

**Comment by parano** [comment]
> This PR address this by assigning the checksum of given bentofile.yaml as a cache key.

does it matter if a referenced requirements.txt file change?

**Pull request merged**